File: Day129_2025/Q129_GFG.py
class Solution:
    def __init__(self):
        self.maxNum = ""

    def findMaximumNum(self, s, k):
        def helper(s_list, k, idx):
            if k == 0:
                return

            max_digit = max(s_list[idx:])

            if max_digit != s_list[idx]:
                k -= 1

            for i in range(len(s_list) - 1, idx - 1, -1):
                if s_list[i] == max_digit:
                    s_list[i], s_list[idx] = s_list[idx], s_list[i]
                    curr = ''.join(s_list)
                    if curr > self.maxNum:
                        self.maxNum = curr
                    helper(s_list, k, idx + 1)
                    s_list[i], s_list[idx] = s_list[idx], s_list[i]  # backtrack

        self.maxNum = s
        helper(list(s), k, 0)
        return self.maxNum

File: Day129_2025/Q129_GFG.java
class Q129_GFG {
    String maxNum;

    private void helper(char[] str, int k, int idx) {
        if (k == 0) return;

        char maxDigit = str[idx];
        for (int i = idx + 1; i < str.length; i++) {
            if (str[i] > maxDigit) {
                maxDigit = str[i];
            }
        }

        if (maxDigit != str[idx]) {
            k--;
        }

        for (int i = str.length - 1; i >= idx; i--) {
            if (str[i] == maxDigit) {
                swap(str, i, idx);
                String current = new String(str);
                if (current.compareTo(maxNum) > 0) {
                    maxNum = current;
                }
                helper(str, k, idx + 1);
                swap(str, i, idx); // backtrack
            }
        }
    }

    private void swap(char[] str, int i, int j) {
        char temp = str[i];
        str[i] = str[j];
        str[j] = temp;
    }

    public String findMaximumNum(String s, int k) {
        maxNum = s;
        helper(s.toCharArray(), k, 0);
        return maxNum;
    }
}
